packages/hierarchy-builder/src/hierarchy-builder/internal/operators/LabelGrouping.ts

```diff
@@ -0,0 +1,112 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import { from, mergeMap, Observable, tap, toArray } from "rxjs";
+import { HierarchyNode } from "../../HierarchyNode";
+import { getLogger } from "../../Logging";
+import { createOperatorLoggingNamespace } from "../Common";
+
+const OPERATOR_NAME = "Grouping.ByLabel";
+/** @internal */
+export const LOGGING_NAMESPACE = createOperatorLoggingNamespace(OPERATOR_NAME);
+
+/** @internal */
+export function createLabelGroupingOperator() {
+  return function (nodes: Observable<HierarchyNode>): Observable<HierarchyNode> {
+    return nodes.pipe(
+      log((n) => `in: ${n.label}`),
+      // need all nodes in one place to group them
+      toArray(),
+      // group all nodes and format into a nodes observable
+      mergeMap((resolvedNodes) => {
+        const obs = from(createLabelGroups(resolvedNodes));
+        return obs;
+      }),
+      log((n) => `out: ${n.label}`),
+    );
+  };
+}
+
+function createLabelGroups(nodes: HierarchyNode[]): HierarchyNode[] {
+  if (nodes.length === 0) {
+    return nodes;
+  }
+  const [firstNode, firstHasChanged] = createLabelGroupsIfClassGroupingNode(nodes[0]);
+  const outputNodes: HierarchyNode[] = [firstNode];
+  let hasChanged = firstHasChanged;
+
+  for (let i = 1; i < nodes.length; ++i) {
+    const [currentNode, currentHasChanged] = createLabelGroupsIfClassGroupingNode(nodes[i]);
+    if (currentHasChanged) {
+      hasChanged = currentHasChanged;
+    }
+
+    if (currentNode.label === outputNodes[outputNodes.length - 1].label) {
+      const lastOutputNode = outputNodes[outputNodes.length - 1];
+      if (HierarchyNode.isLabelGroupingNode(lastOutputNode) && Array.isArray(lastOutputNode.children)) {
+        if (currentNode.params?.groupByLabel) {
+          lastOutputNode.children.push(currentNode);
+        } else {
+          outputNodes.push(currentNode);
+          [outputNodes[outputNodes.length - 1], outputNodes[outputNodes.length - 2]] = [
+            outputNodes[outputNodes.length - 2],
+            outputNodes[outputNodes.length - 1],
+          ];
+        }
+      } else if (lastOutputNode.params?.groupByLabel) {
+        if (currentNode.params?.groupByLabel) {
+          outputNodes[outputNodes.length - 1] = {
+            label: currentNode.label,
+            key: {
+              type: "label-grouping",
+              label: currentNode.label,
+            },
+            children: [lastOutputNode, currentNode],
+          };
+        } else {
+          outputNodes.push(currentNode);
+          [outputNodes[outputNodes.length - 1], outputNodes[outputNodes.length - 2]] = [
+            outputNodes[outputNodes.length - 2],
+            outputNodes[outputNodes.length - 1],
+          ];
+        }
+      } else {
+        outputNodes.push(currentNode);
+      }
+    } else {
+      outputNodes.push(nodes[i]);
+    }
```

Comment:
I believe having one `outputNodes.push(currentNode)` just before ending the loop an a couple of `continue` in cases we don't need to call it would be cleaner.

Reply:
Changed